### PR TITLE
Add JujuVersion.has_secrets()

### PR DESCRIPTION
**Pull request by beliaev-maksim**
add logic to check if secrets supported

```python
if juju_version.secrets_supported:
    # do smth with secrets
```

### CODE REVIEW

**Comment by jnsgruk** [comment]
I'm not sure this is necessarily the right approach. We already have `assumes` that can be used in `metadata.yaml` to specify a Juju version, and we could probably add a `secrets` entry (@wallyworld?).

The secrets implementation (currently WIP) can just raise an exception if the `secret-*` hook tool being called is not present in the hook tools dir, allowing charm authors to catch that case and handle things differently?

**Comment by beliaev-maksim** [comment]
@jnsgruk 
if I got `assumes` right, then we cannot create a charm that will work on both 2.x and 3.x
while providing a logic to check, we can branch decision

also, doing if/else check looks to be a more cleaner alternative and to my knowledge faster than try/except block in scenarios when except is called

but if we already have a good way to do a check, then we just might need document it instead of modifying the code

**Comment by jnsgruk** [comment]
> if I got `assumes` right, then we cannot create a charm that will work on both 2.x and 3.x
> while providing a logic to check, we can branch decision

Yeh, you're right there. If you use `assumes: ["secrets"]` then the charm would only deploy on `3.x`

As for whether to provide a simple check method or rely on `try`/`except`, I'll defer to @benhoyt for his preference there :+1:

**Comment by jameinel** [comment]
Historically if you want your charm to support versions of juju that both do and don't support a feature, then you would probe to see if the command was available (LookPath secret-get, etc). That was used when we introduced leadership and 'is-leader'.
And then you would want branching code paths in your charm.

However, if you don't want branching code paths, and you just want to move to secrets always, then Juju should definitely have an `assumes: [secrets]`, I'm not sure if we have that with 3.0.0, but we should definitely add it in 3.0.
That would prevent 2.9 from deploying the charm, and the charm just always assumes that secret-get is available.

The harder problem (IMO), is more about relation interface versioning. If the new version of postgresql wants to use secrets to share its database password, it needs to know whether Juju supports it (but as above that is relatively obvious), but it also needs to know if the *related application* knows how to interact with secrets.
And that is more a relation-interface versioning problem. (eg, what is the actual protocol of `interface: pgsql`). We've had lots of conversations around versioning, and how to maintain compatibility, etc. It gets quite tricky because if postgresql wants to start using secrets, it can put that information in the relation data, but wordpress will read the relation data, and not find a password in the relation, and assume it just needs to wait until a password shows up, without sending a message back to postgresql that "I don't know what this `password-secret:` field means". 
Which means that the *compatible* relation interface is to still fill a plaintext password into relation data, and a password secret identifier that isn't yet helping.
(or you introduce version negotiation into the relation data, or this pushes us to actually do relation interface versioning in Juju.)

File: ops/jujuversion.py [hunk 1]
```diff
@@ -112,3 +112,8 @@ def is_dispatch_aware(self) -> bool:
     def has_controller_storage(self) -> bool:
         """Determine whether this juju version supports controller-side storage."""
         return (self.major, self.minor, self.patch) >= (2, 8, 0)
+
+    @property
+    def secrets_supported(self) -> bool:
```
Comment (benhoyt) [inline ops/jujuversion.py:117]: I suggest we call this `has_secrets` to be consistent with the other names here:

```suggestion
    def has_secrets(self) -> bool:
```

File: ops/jujuversion.py [hunk 2]
```diff
@@ -112,3 +112,8 @@ def is_dispatch_aware(self) -> bool:
...
+        """Determine whether this juju version supports `secrets` feature."""
+        return (self.major, self.minor, self.patch) >= (3, 0, 0)
```
Comment (benhoyt) [inline ops/jujuversion.py:119]: Let's make this 3.0.2 as that's the first version with the final version of secrets with the `--refresh` arg and `secret-get-info` separate hook command.

```suggestion
        return (self.major, self.minor, self.patch) >= (3, 0, 2)
```

File: ops/jujuversion.py [hunk 3]
```diff
@@ -112,3 +112,8 @@ def is_dispatch_aware(self) -> bool:
     def has_controller_storage(self) -> bool:
         """Determine whether this juju version supports controller-side storage."""
         return (self.major, self.minor, self.patch) >= (2, 8, 0)
+
+    @property
+    def secrets_supported(self) -> bool:
+        """Determine whether this juju version supports `secrets` feature."""
```
Comment (benhoyt) [inline ops/jujuversion.py:118]:
```suggestion
        """Determine whether this juju version supports the `secrets` feature."""
```

**Review by benhoyt — CHANGES_REQUESTED** [review]
I'll add further comments separately, but we also need unit tests for this, even though it's a trivial feature.

**Comment by benhoyt** [comment]
I think this is actually a reasonable addition, given that this would be useful for folks that want to create charms that work on 2.9 and 3.0 (without secrets or with secrets), and we have some precedent for these helpers (`has_controller_storage` and others). It seems nicer for the charmer than having to explicitly know to do something like `shutil.which('secret-get')`.

As @jameinel mentioned, if you just want to say "this charm *requires* secrets", `assumes` seems like the right way to go. We don't have an `assumes: [secrets]` right now; you'd have to say `assumes: [juju >= 3.0]` or more precisely `assumes: [juju >= 3.0.2]`. Is that enough, given that it's basically 2.9 vs 3.0 anyway?

> The harder problem (IMO), is more about relation interface versioning.

Yes, that's definitely true. Though that's an existing problem with any relation data and should probably be considered separately.

**Comment by PietroPasotti** [comment]
Let's not do relation versioning in the databag -- got PTSD from the last shot we took at that.

I'm on board with adding `secrets_supported`, and this also makes me realize that we probably need something like a `ops.testing.SECRETS: bool` global that we can use to toggle whether the harness has secret simulation enabled or not.
Specifically for the case where one wants to test a charm that is compatible both with secret-less and newer juju versions.

Precedent for that pattern would be `testing.SIMULATE_CAN_CONNECT`, or perhaps we need a broader abstraction of the api version we're mocking against. @benhoyt is this something I should be writing also in the ops secrets spec?

**Comment by benhoyt** [comment]
> Precedent for that pattern would be testing.SIMULATE_CAN_CONNECT, or perhaps we need a broader abstraction of the api version we're mocking against. @benhoyt is this something I should be writing also in the ops secrets spec?

Ops secrets spec -- I wouldn't worry about adding it to that at this point. I'll think about `ops.testing.SECRETS` type of thing as I work though the secrets code.

**Review by jnsgruk — APPROVED** [review]
This looks simple enough, and inline with similar checks that already exist. Thanks for the iteration!

**Review by benhoyt — APPROVED** [review]

**Comment by benhoyt** [comment]
@beliaev-maksim BTW, in future it's probably best to have your `@canonical.com` email address to show up on these commits.

**Pull request merged**